File: solutions/csharp/word-count/1/WordCount.cs
using System.Text.RegularExpressions;

public static class WordCount
{
    public static IDictionary<string, int> CountWords(string phrase)
    {
        return Regex
            .Matches(phrase.ToLower(), @"\b\w+(?:'\w+)?\b")
            .Select(match => match.Value)
            .GroupBy(word => word)
            .ToDictionary(group => group.Key, group => group.Count());
    }
}

File: solutions/fsharp/word-count/1/WordCount.fs
module WordCount

open System.Text.RegularExpressions

let private incrementCount counts word =
    let count = counts |> Map.tryFind word |> Option.defaultValue 0
    Map.add word (count + 1) counts

let countWords (phrase: string): Map<string, int> =
    Regex.Matches(phrase.ToLowerInvariant(), "[a-z0-9]+(?:'[a-z0-9]+)?")
    |> Seq.cast<Match>
    |> Seq.map (fun matched -> matched.Value)
    |> Seq.fold incrementCount Map.empty
